feat(AVM)!: Check start and end states in the tx trace

sirasistant opened the pull request:
Checks initial and end state from public inputs, reusing as many lookups as possible at the end. Enables many disabled lookups now that the whole bulk test is simulated

dbanks12 commented on barretenberg/cpp/pil/vm2/tx.pil:
```diff
@@ -33,9 +33,13 @@ namespace tx;
+
+    pol commit start_tx;
+    #[START_WITH_SEL]
+    (1 - sel) * sel' * (1 - start_tx') = 0;
```
Does anything prevent `start_tx` from being 1 on other rows too?

sirasistant replied:
hmmm I think i messed up that one

dbanks12 commented on barretenberg/cpp/pil/vm2/tx_context.pil:
```diff
@@ -23,6 +23,139 @@ namespace tx;
+        precomputed.clk,
+        public_inputs.cols[0],
+        public_inputs.cols[1]
+    };
```
What exactly is missing that would let us remove this TODO?

sirasistant replied:
this is old, good catch, before I added the final phase!

dbanks12 commented on barretenberg/cpp/pil/vm2/tx_context.pil:
```diff
@@ -23,6 +23,139 @@ namespace tx;
+        public_inputs.cols[0],
+        public_inputs.cols[1]
+    };
+
+    start_tx * (constants.AVM_WRITTEN_PUBLIC_DATA_SLOTS_TREE_INITIAL_ROOT - prev_written_public_data_slots_tree_root) = 0;
+    start_tx * (constants.AVM_WRITTEN_PUBLIC_DATA_SLOTS_TREE_INITIAL_ROOT - prev_written_public_data_slots_tree_root) = 0;
+
```
I don't see where exactly this is used. The other start_tx constant constraints are used right afterwards, like "set some PI offset, then do the PI lookup". What do these constraints accomplish?

Basically, the "written slots" tree is a bit less intuitive to me. I know that it isn't a normal state tree, but it might be worth a comment here just saying "this tree is an AVM-internal tree for keeping track of written public data slots. We given its root an initial value at the start of the TX"

sirasistant replied:
will do!

dbanks12 commented on barretenberg/cpp/pil/vm2/tx_context.pil:
```diff
@@ -79,40 +213,49 @@ namespace tx;
+    #[PROPAGATE_DA_GAS_USED]
+    NOT_LAST_ROW * (next_da_gas_used - prev_da_gas_used') = 0;
+
+    #[PROPAGATE_L2_GAS_LIMIT]
+    NOT_LAST_ROW * (1 - is_teardown_phase') * (l2_gas_limit - l2_gas_limit') = 0;
+    #[PROPAGATE_DA_GAS_LIMIT]
+    NOT_LAST_ROW * (1 - is_teardown_phase') * (da_gas_limit - da_gas_limit') = 0;
```
I guess the gas limit comes from a PI lookup for teardown. It's worth mentioning that here in a comment and maybe even calling out that lookup's label `PUBLIC_INPUTS_READ_GAS_LIMIT` since that lookup is not grouped near this constraint.

dbanks12 commented on barretenberg/cpp/pil/vm2/tx_context.pil:
I'd throw an additional comment in per-PI write. Just a one-liner like "READ START/END STATE FROM PI FOR NULLIFIERS". If nothing else, it'll just visually break up the file a bit more.

dbanks12 commented on barretenberg/cpp/src/barretenberg/vm2/common/aztec_types.hpp:
```diff
@@ -27,6 +27,8 @@ enum TransactionPhase {
     APP_LOGIC = 8,
     TEARDOWN = 9,
     COLLECT_GAS_FEES = 10,
+    TREE_PADDING = 11,
+    CLEANUP = 12,
 };
```
We really need a good doc explaining all of the responsibilities of each phase, especially since these are really circuit concepts that don't exist in the TS simulator at all. I'm not saying that needs to be done in this PR, but we'll need it fairly soon.

sirasistant replied:
Will do next week!

dbanks12 commented on yarn-project/simulator/src/public/avm/opcodes/external_calls.ts:
```diff
@@ -147,6 +147,10 @@ export class SuccessCopy extends Instruction {
     const memory = context.machineState.memory;
     const addressing = Addressing.fromWire(this.indirect);
 
+    context.machineState.consumeGas(
+      this.baseGasCost(addressing.indirectOperandsCount(), addressing.relativeOperandsCount()),
+    );
+
     const operands = [this.dstOffset];
```
Yikes.... Was this just a bug?

sirasistant replied:
yep!

dbanks12 reviewed (approved):
LGTM! Just some nits requesting comments

sirasistant committed:
feat(AVM)!: Check start and end states in the tx trace

Checks initial and end state from public inputs, reusing as many lookups as possible at the end. Enables many disabled lookups now that the whole bulk test is simulated
